[DOCS] Reword node roles docs

jrodewig opened the pull request:
Rewords the node roles introduction to better emphasize that node roles much be explicitly set.
Also de-emphasizes the default roles.

### Preview
https://elasticsearch_69301.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/modules-node.html#node-roles

jrodewig committed:
[DOCS] Reword node roles intro para

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

elasticmachine commented:
Pinging @elastic/es-docs (Team:Docs)

jrodewig committed:
Rewrap lines

sophiec20 commented:
Can we go further? 

For example removing the mention of `(default)` in the separate paragraphs below, and especially removing `which is the default behavior in the Elasticsearch default distribution` from the ML section.

From troubleshooting the resultant impact of multiple issues caused by migrating to using this setting, reading this page did not immediately help. Perhaps a leading high impact call-out -- something along the lines of `Important: If you are migrating to using this setting, ensure you have every node role specified that you will need. A basic cluster requires at least one master, data, ingest, transform node.` (although check this list is enough).

jrodewig committed:
Address review feedback

jrodewig commented:
Thanks for the additional feedback @sophiec20. I've removed the `(default)` mentions as well as some outdated references to the default distribution.  I also added the admonition you suggested with an explicit example for transform nodes.

Please let me know if you have any additional thoughts! Thanks!

sophiec20 commented on docs/reference/modules/node.asciidoc:
```diff
@@ -32,42 +33,45 @@ configure this setting, the node has the following roles by default:
+cluster needs. For example, to use {transforms}, you must specify the
+`transform` role for one or more nodes in your cluster. For resiliency,
+production {es} clusters require an elected `master` node and at least one node
+for each role. See <<high-availability-cluster-design>>.
```
"For resiliency" we will need more than "at least one node for each role". Resiliency requires redundancy, so this final sentence isn't quite right however the link to the HA cluster page is useful.

I personally prefer the concept of naming which roles are most likely to be needed. For example `...A minimal cluster requires at least one master, data, ingest and transform node. For production, resilience is required ...`.

The reason why I like naming the roles, is that a user may not know they are using a transform, because it was installed as part of endpoint. They may have no requirement to use transforms as far as they are aware, but are still using them.

jrodewig replied:
:doh: Thanks for catching my error. I obviously got my requirements mixed up. 😬  Now that I better understand your feedback, I'm not sure if we should include that section anyway.

I think the definition of a `minimal` cluster will likely differ based on use case, but I agree that we should highlight which roles are needed to access most Stack features. 

I've revised the admonition based on your suggestion and tested locally to ensure our solutions still stay up. As you mentioned offline, I also feel that this should be addressed in the UI.

As this now falls into the area of a "best practice," I'll also get a review from @jasontedor. Thanks again, @sophiec20!

jrodewig committed:
Revise admonition

jrodewig committed:
Remove sentence

jrodewig committed:
Remove unneeded sentence

jrodewig commented on docs/reference/modules/node.asciidoc:
```diff
@@ -32,42 +33,52 @@ configure this setting, the node has the following roles by default:
 * `transform`
 
-NOTE: If you set `node.roles`, the node is assigned only the roles you specify.
+[IMPORTANT]
+====
+IMPORTANT: If you set `node.roles`, ensure you specify every node role your
+cluster needs. To access most {stack} features, a minimal cluster requires nodes
+with the following roles:
+
+- `master`
+- `ingest`
+- `transform`
+- `data`, `data_content`, or a combination of the `data_hot`, `data_warm`,
+  `data_cold`, and `data_frozen` roles
+====
```
@jasontedor Are you comfortable with including this as a best practice?

The goal is to enable users to access most Stack features, such as Endpoint, if they set `node.roles`. The chief concern is that a user may not be aware that Endpoint or similar solutions require these roles otherwise.

Any other feedback you have is appreciated. Thanks!

jrodewig committed:
Rewrap lines

sophiec20 reviewed (approved):
LGTM 
Subject to additional best practice review.

jasontedor commented on docs/reference/modules/node.asciidoc:
```diff
@@ -32,42 +33,52 @@ configure this setting, the node has the following roles by default:
+- `master`
+- `ingest`
+- `transform`
+- `data`, `data_content`, or a combination of the `data_hot`, `data_warm`,
```
`data_content` should be part of second clause in the disjunction. A node either has the `data` role or it has specific `data_` roles. `data_content` is another tier in the data tiering.

jasontedor commented on docs/reference/modules/node.asciidoc:
```diff
@@ -32,42 +33,52 @@ configure this setting, the node has the following roles by default:
+
+- `master`
+- `ingest`
+- `transform`
```
I'm curious why `transform` but not `ml`? And what do we consider "most" Stack features? For example, is CCS here, or is that seen as advanced? If it's under the umbrella of "most" Stack features, we need `remote_cluster_client` in this list too. And now we're basically listing every role. Either way, we have almost every role listed here but not much guidance how a user decides which roles they need. I wonder if we should present it that way: use the `X` role if you're trying to do `Y` (e.g., `remote_cluster_client` if you're trying to search or replicate from another cluster, `ml` if you're trying to do analomy detection (or use `{ml-features}`), etc.)?

jrodewig replied:
Thanks for the suggestion @jasontedor. I went with this approach, which made the admonition simpler.

Rather than include a `minimal` list of roles. This now just provides guidance on the roles required for specific features.

sophiec20 replied:
From reading the whole page, it seemed to me the it is written in the form "Use `X` role if you are trying to do `Y`". This is already covered.

The additional call-out box was to highlight the most obvious "getting started" roles by name. This advice could save users some time as they will want to ingest data and there is a transform that is deployed with endpoint. The user did not actively choose to load the transform and so they may not realise that they will need this role.  

It might be difficult for us to come together and agree what this initial typical set of roles would be. 

However I still think we can use our docs to anticipate problems for users who are deciding to use `node.roles`. They may be migrating from using a legacy setup where their data nodes could run transforms without additional configuration. They may have intentionally created transforms or they may just have the endpoint transform enabled.

In which case, instead of the list of typical roles, we could limit the call-out to only mention that the transform role must now be explicitly set if either the user or the Security Solution has been created one. Happy to do either, as both options help to avoid a likely configuration problem.

jrodewig replied:
Thanks @sophiec20. While the admon isn't limited to just the `transform` role, it does now highlight that you need to set the `transform` role for the Elastic Security app, which wasn't highlighted previously. This should prevent users from accidentally removing a role they didn't think was required (because they didn't know Elastic Security used transforms).

sophiec20 replied:
@kevinlog Please confirm when the endpoint transform is created. It gets installed with Fleet. Are there other conditions under which it is installed?

kevinlog replied:
@sophiec20 it gets installed whenever the Endpoint Package is installed.  This happens in 2 places.

1. When navigating to Fleet as a superuser as you mentioned
2. When navigating to the Security application as a superuser

Those are the only conditions when it is installed

jrodewig replied:
Thanks @sophiec20 and @kevinlog. I've noted that Fleet also requires the `transform` node role.

sophiec20 replied:
Thx LGTM

jasontedor reviewed (commented):
I left a few comments.

jrodewig committed:
Add per-feature role section

jasontedor reviewed (approved):
LGTM. Thanks @jrodewig!

jrodewig committed:
Add fleet to `transform` node features

jrodewig committed:
Fix typo
